Add support for listing and launching emulators

DanTup opened the pull request:
This is work-in-progress, just opening this to allow feedback.

For now, I'm using the term `emulators` to refer to these (these being AVDs in the case of Android and "the simulator" for iOS), but this can easily be changed.

Commands are being implemented as:

- `flutter emulators` - lists emulators
- `flutter emulators --launch xxx` - launch an emulator

Creation will be handled in a separate PR once listing/launching is done. The code is mostly a copy of the `devices` code but much simplified (since there's a lot less to do with an AVD). Right now, this just supports `flutter emulators` which spits out the list (same as `emulator -list-avds`).

## TODO:

- [x] Display more information about an emulator (currently only shows IDs)
- [x] Support launching an emulator
- [x] [Are there other places the simulator could be](https://github.com/flutter/flutter/pull/16705#discussion_r185809497)?
- [x] [Fix missing await](https://github.com/flutter/flutter/pull/16705#issuecomment-386339354) (in a way that doesn't stall the process until emulator closes, or lose error output)
- [ ] Add tests
- [x] Agree on name (currently "emulators")
- [x] Agree command interface (currently `flutter emulators` w/ `--launch`/`--create`)
- [x] Ensure missing `emulators` command reports a useful error
- [x] Ensure error output from `emulators` command is displayed correctly
- [x] Add in "the simulator" for iOS when running on Mac
- [x] Ensure iOS simulator doesn't appear on non-Mac
- [x] ~~Investigate if ID/names can be matched up with `flutter devices` output~~ Yes they can, but only be connected to the device. Need to decide whether to do anything with this (considering it out of scope for this PR)

Relates to:

https://github.com/flutter/flutter/issues/14822
https://github.com/Dart-Code/Dart-Code/issues/490
https://github.com/flutter/flutter/issues/13379

DanTup committed:
Add basic support for listing Android AVDs

Very basic support for "flutter emulators" which just lists the available Android AVDs.

Relates to:

https://github.com/flutter/flutter/issues/14822
https://github.com/Dart-Code/Dart-Code/issues/490
https://github.com/flutter/flutter/issues/13379

DanTup commented:
@devoncarew @mit-mit I've been working through the Android Studio code to see how it gets more info about AVDs and it seems that it doesn't just shell out to `emulator` as I am, it basically parses the `AVD` ini files itself, finding the location like:

```java
    public static String getAvdFolder() throws AndroidLocationException {
        if (sAvdLocation == null) {
            String home = findValidPath(Global.ANDROID_AVD_HOME);
            if (home == null) {
                home = getFolder() + FOLDER_AVD;
            }
            sAvdLocation = home;
            if (!sAvdLocation.endsWith(File.separator)) {
                sAvdLocation += File.separator;
            }
        }
        return sAvdLocation;
    }
```

Do we want to copy this? Without it, I think we could only display the IDs so I think we do, but it's a big step from just parsing the output of `emulator` as we need to start coding in fairly specific knowledge about how emulator etc. works (I don't know if it's likely to change, but it's a risk).

DanTup committed:
Fix references to Devices and remove unused properties

mit-mit commented:
Yes, we would very much like to show the AVD names, this is https://github.com/flutter/flutter/issues/11850

DanTup committed:
Add some emulator tests

DanTup commented:
> Yes, we would very much like to show the AVD names, this is #11850

That might be the opposite problem - that's getting the AVD names from a running device, whereas this is about getting more info on the AVDs. Matching up a connected device to an AVD ini on disk might be more complicated (I might have a better idea once I've done dug into the ini files for the AVD side).

DanTup committed:
Read information about AVDs from config.ini

DanTup committed:
Add support for launching emulators

DanTup committed:
Add some tests around device+ini parsing

DanTup commented:
The IDs in `flutter devices` seem totally different - I'm not sure if we can match them up (I presume the info that `flutter devices` is getting is runtime info from the device, whereas the AVD info is all just metadata and user-defined names?).

<img width="685" alt="screen shot 2018-04-18 at 14 52 12" src="https://user-images.githubusercontent.com/1078012/38937309-d6107e76-431a-11e8-9fa1-6aa04ecb90a0.png">

DanTup committed:
Add support for iOS Simulator in `flutter emulators`

DanTup committed:
Improve formatting

DanTup committed:
Fix typo

DanTup committed:
Remove launch status

It's too fast and leaves weird output on the screen

DanTup committed:
Fix crash on Windows caused by no HOME var

Windows sets HOMEDRIVE/HOMEPATH.

DanTup committed:
Tidy up

DanTup committed:
Improve error message when no AVD path

DanTup committed:
Don't try to build emulators from empty output

DanTup committed:
Handle nulls caused by missing information

DanTup committed:
More Windows fixes

Added an exists check to ensure we don't try to run if emulator is missing, but that requires the file extension for Windows.

DanTup commented:
@devoncarew @mit-mit It might not be bullet proof yet, but this generally woks for listing and launching Android and iOS emulators on my Mac and Windows now:

<img width="1278" alt="screen shot 2018-04-18 at 15 42 27" src="https://user-images.githubusercontent.com/1078012/38944167-12532d4c-432b-11e8-9f00-84cab5d9bf2b.png">

I've added some basic tests for the emulator classes (sorta copied from the equiv with devices) and the ini parsing; I don't know if we could easily have end-to-end tests here (do our CIs have Android SDKs and AVDs we could use without having to mock the files on disk etc.?).

Feedback appreciated!

devoncarew commented:
> it basically parses the AVD ini files itself, finding the location like

Do you have a sense for how stable that is, Android SDK release to release? That's less of a stable API surface area than the `emulator` command.

> do our CIs have Android SDKs and AVDs we could use without having to mock the files on disk etc

Our devicelabs bots very likely do.

devoncarew commented on packages/flutter_tools/lib/src/android/android_emulator.dart:
```diff
@@ -0,0 +1,115 @@
+  bool get supportsPlatform => true;
+
+  @override
+  bool get canListAnything => androidWorkflow.canListEmulators;
```
I think we can remove this method for the emulators - this is used for devices to know whether we should display an error in flutter doctor; we won't need that for emulators.

devoncarew replied:
Ah, I see you use it below.

DanTup replied:
Yeah, if we can't find `emulator` or the AVD folder, we output a message.

devoncarew commented on packages/flutter_tools/lib/src/android/android_emulator.dart:
```diff
@@ -0,0 +1,115 @@
+        await runAsync(<String>[getEmulatorPath(), '-avd', id]);
+    if (launchResult.exitCode != 0) {
+      printError(
+          'Error: emulator exited with exit code ${launchResult.exitCode}');
```
It may not be necessary to print this explicitly - the `printError('$launchResult')` will print the stdout+stderr.

DanTup replied:
Yep, message seems fine without the first line so I've removed it (the original code was copied from `android_device.dart` for ADB install, but it seems like that's the only place doing that!).

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -0,0 +1,82 @@
+  final String name = 'emulators';
+
+  @override
+  final String description = 'List all available emulators.';
```
`List and start available emulators.`?

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -0,0 +1,82 @@
+        await emulatorManager.getEmulatorsById(id).toList();
+
+    if (emulators.isEmpty) {
+      printStatus("No emulator found that matches the ID '$id'.");
```
Do emulators have IDs? If no, perhaps call this an emulator name.

devoncarew replied:
Ah, something like `No emulator found matching '$id'.` ?

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -0,0 +1,82 @@
+      printStatus("More than one emulator matches the ID '$id':\n");
+      Emulator.printEmulators(emulators);
+    } else {
+      emulators.first.launch();
```
`await` this?

DanTup replied:
Ugh, looks like there was a bug here. When I await it, the process never terminates until the emulator is closed. I'm not sure how to fix this - we have a `runDetached` method but then we wouldn't get any output. Any suggestions?

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -0,0 +1,82 @@
+      printStatus('No emulators available.\n\n'
+          // TODO: Change these when we support creation
+          // 'You may need to create images using "flutter emulators --create"\n'
+          'You may need to create one using Android Studio\n'
```
nit: remove the newline?

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -0,0 +1,82 @@
+
+    if (emulators.isEmpty) {
+      printStatus('No emulators available.\n\n'
+          // TODO: Change these when we support creation
```
`TODO(dantup): ...`

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -0,0 +1,82 @@
+      if (diagnostics.isNotEmpty) {
+        printStatus('');
+        for (String diagnostic in diagnostics) {
+          printStatus('• ${diagnostic.replaceAll('\n', '\n  ')}');
```
I don't think we prefix lists of errors w/ bullets elsewhere. If we know these are list items, perhaps indenting would be enough; else we can just print the lines.

DanTup replied:
This was actually copied from the devices file. We don't currently use it; I wasn't sure if to leave it in case we add some or remove. I'll remove for now, since I'm not sure we'll really have any good use for it.

devoncarew commented on packages/flutter_tools/lib/src/ios/ios_emulators.dart:
```diff
@@ -0,0 +1,73 @@
+import '../globals.dart';
+import 'ios_workflow.dart';
+
+// TODO: Is there a better name for this? We already have IOSSimulator classes
```
IOSSimulatorManager?

devoncarew commented on packages/flutter_tools/lib/src/emulator.dart:
```diff
@@ -0,0 +1,176 @@
+
+  /// A user-specified emulator ID.
+  String get specifiedEmulatorId {
+    if (_specifiedEmulatorId == null || _specifiedEmulatorId == 'all')
```
This would launch all emulators? I'm not sure we need to support that. If we do, we could special case in EmulatorCommand.

DanTup replied:
Whoops, this was unused code. It was copied from devices and I didn't come back to it. I'll remove, but also implement the ability to filter the list in the same way that this was used for in devices.

devoncarew commented on packages/flutter_tools/lib/src/emulator.dart:
```diff
@@ -0,0 +1,176 @@
+    return _emulatorDiscoverers.where((EmulatorDiscovery discoverer) => discoverer.supportsPlatform);
+  }
+
+  /// Return the list of all connected emulators.
```
==> available emulators

devoncarew commented on packages/flutter_tools/lib/src/emulator.dart:
```diff
@@ -0,0 +1,176 @@
+  @override
+  String toString() => name;
+
+  static Stream<String> descriptions(List<Emulator> emulators) async* {
```
I'd just convert this to a `List<String>` for simplicity.

devoncarew commented on packages/flutter_tools/lib/src/emulator.dart:
```diff
@@ -0,0 +1,176 @@
+    final List<List<String>> table = <List<String>>[];
+    for (Emulator emulator in emulators) {
+      table.add(<String>[
+        emulator.name ?? emulator.id ?? '',
```
It looks like we always have IDs, but not always names. Perhaps:

id (name) • manufacturer • label

What is label (vs name)?

DanTup replied:
I thought label could be used for additional info that might vary between platforms (eg. architecture or other stuff we might want to build into a string but not give its own column as it's not consistent between Android/iOS).

The ID is the unique thing a user can use to launch an emulator regardless of how unique the name is, so I think it should be in its own column (however it does look a bit naff).

If we put `id (name)` in the first column, it'd could look like `Nexus_5X_API_27_X86 (Nexus 5X)` which seems a bit of a moutful. WDYT?

DanTup replied:
Ugh, apparently Cmd+Enter starts a review! Why would I want to review my own changes? 😆

DanTup replied:
I've gone with the first for now, so ID is the first column. We can always tweak it further if we think it looks weird with similar name/IDs.

devoncarew replied:
👍

devoncarew commented on packages/flutter_tools/lib/src/ios/ios_emulators.dart:
```diff
@@ -0,0 +1,73 @@
+  String get manufacturer => 'Apple';
+
+  @override
+  String get label => '';
```
return null? It looks like `descriptions()` handles that

devoncarew commented on packages/flutter_tools/lib/src/emulator.dart:
```diff
@@ -0,0 +1,176 @@
+    final Emulator exactMatch = emulators.firstWhere(
+        exactlyMatchesEmulatorId, orElse: () => null);
+    if (exactMatch != null) {
+      yield exactMatch;
```
I tend not to use `yield`s myself, but it does look like we use it elsewhere in flutter_tools.

DanTup replied:
Yeah, it felt a bit weird to me tbh (especially as we call one function, toList it, then filter again), was just how it was in the code I copied. I might rip it out though, I don't think there's any benefit to having an iterable here (we're `toList()`ing it further up anyway).

DanTup replied:
Actually, turns out this is because we enumerator multiple discovered and yield the results (to avoid adding to a list). We don't need to do it in `getEmulatorsById` (which I've renamed to `getEmulatorsMatching` since it isn't strictly ID) however it might br weird if `getAllAvailableEmulators` and `getEmulatorsMatching` returned different types (Stream vs List).

So, if you have any opinions on this, I'm happy to change, but leaving as-is for now.

devoncarew replied:
It's not important to change for this PR, but you should be able to change the return type to `Future<List<Emulator>>`, return `[exactMatch]` above, and return a filtered view of the emulators below.

DanTup replied:
Yeah, that's what this comment was about:

> however it might br weird if `getAllAvailableEmulators` and `getEmulatorsMatching` returned different types (`Stream` vs `List`).

Making `getEmulatorsMatching` return ` Future<List>` is simple, but I can't figure out how to change `getAllAvailableEmulators` to do the same to match:

```dart
  Stream<Emulator> getAllAvailableEmulators() async* {
    for (EmulatorDiscovery discoverer in _platformDiscoverers) {
      for (Emulator emulator in await discoverer.emulators) {
        yield emulator;
      }
    }
  }
```

I can do `await getAllAvailableEmulators().toList()` and that does it, but I can't figure out how to fix the method itself. It seems like there should be a simple way (without just putting it inside a nested function) to do that from within?

devoncarew replied:
something perhaps like:

```
Future<List<Emulator>> getAllAvailableEmulators() async {
  List<Emulator> emulators = <Emulator>[];
  for (EmulatorDiscovery discoverer in _platformDiscoverers) {
    for (Emulator emulator in await discoverer.emulators) {
        emulators.add(emulator);
    }
  }

return emulators;
}
```

It's simple; it's possible you could also use `Future.wait` for slightly more terse code.

DanTup replied:
Heh, I thought there might be a cleaner way. I got a lint(?) on your code saying to use `.forEach`, which I had to swap for `Future.forEach` to get it to work; so I ended up with:

```dart
  /// Return the list of all available emulators.
  Future<List<Emulator>> getAllAvailableEmulators() async {
    final List<Emulator> emulators = <Emulator>[];
    Future.forEach(_platformDiscoverers, (EmulatorDiscovery discoverer) async {
      emulators.addAll(await discoverer.emulators);
    });
    return emulators;
  }
```

I'll find a way to eliminate that list without making the code unreadable one day!

devoncarew reviewed (commented):
Fantastic! I made a quick pass through.

DanTup commented:
> Do you have a sense for how stable that is, Android SDK release to release? That's less of a stable API surface area than the `emulator` command.

Not really. I'd guess they're relatively stable since there are many tools (and versions of tools) that read them, but it is a risk. I do believe I handle when we can't find the AVD folder (displaying a message) or the config/ini but I don't know if there are any options besides not having the extra info or taking this risk.

FWIW, the original hope was that the names would match up with `flutter devices` and they don't, so we're really just getting a prettier display name and some other info (we can see screen res, memory, etc.) we could display. If we don't think it's worth it, we can remove this (though we'd just have IDs like `Nexus_5X_x64_API_27` which may be a little unfriendly).


> Our devicelabs bots very likely do.

Are they likely to have a consistent environment all with the same AVDs set up? I don't think I can check this without merging test code into master and logging it. Even they do, maybe it'd make things a little fragile - maybe it'd be better once we add `create` to add tests here, so we can create them, make sure they're listed, maybe even launch them.

DanTup committed:
Remove redundant lines from errors

DanTup committed:
Improve text

DanTup committed:
Tweak text because we check IDs and names

And it's not obvious to the use what the ID is.

DanTup committed:
Remove newline

DanTup committed:
Add dantup to TODO comments

DanTup committed:
Remove unused emulator diagnostics

DanTup committed:
Remove unused code

DanTup committed:
Fix filtering of emulator list

`flutter emulators nexus` will now return only those that'd match (eg. same things that `flutter emulators --start nexus` would match).

DanTup committed:
Fix typo

DanTup committed:
Flatten descriptions from Stream to List

DanTup committed:
Use null instead of empty

DanTup commented:
@devoncarew Thanks! I think I've addressed all the comments except:

- Issues with awaiting the launch of `emulator`
- Use of yield in `getEmulatorsMatching`

And the above comments about integration tests.

mit-mit commented:
@DanTup I have no idea how it does it, but when you hit Run in Android Studio and get prompted to select a target, it does seem to be able to list the running device with the same information/IDs as in the AVD dialog:

![screen shot 2018-04-19 at 11 06 33](https://user-images.githubusercontent.com/13644170/38982344-e18fb92c-43c1-11e8-82e5-5ae1a576313c.png)

mit-mit commented:
> For now, I'm using the term emulators to refer to these (these being AVDs in the case of Android and "the simulator" for iOS), but this can easily be changed.

I wonder if it would be better to just include these in `flutter devices`, e.g.:

```
$ flutter devices
2 connected, running devices:

Android SDK built for x86 • emulator-5554                        • android-x86 • Android 8.0.0 (API 26) (emulator)
iPhone X                  • 6545ED50-3F15-479F-937E-CB0851E43D57 • ios         • iOS 11.3 (simulator)

1 stopped emulator (use flutter devices --start <ID> to start):

Nexus 5X • emulator-5556                        • android-x86 • Android 8.0.0 (API 26) (emulator)

```

DanTup commented:
> it does seem to be able to list the running device with the same information/IDs as in the AVD dialog

Oh yeah, that does look promising - I didn't realise it was categorising the online ones (I didn't see that dialog). I've got the code checked out, so I'll dig through and see what they're doing :-)

> I wonder if it would be better to just include these in `flutter devices`, e.g.:

Possibly - it was one of the options in the doc. One of the things that put us off was not being able to tie them together (which based on the above may have been misguided), which would've resulted in the same things appearing both as online devices and offline. Let me dig into this and see what I can do, and if we can tie them together we can discuss again (I don't have strong opinions as long as we don't confuse the user with the same device listed twice with different names).

DanTup commented:
Ok, I managed to find this (sorry for screenshot, not sure where I can link to this code online):

<img width="930" alt="screen shot 2018-04-19 at 12 56 22" src="https://user-images.githubusercontent.com/1078012/38990191-2c9ea0ae-43d1-11e8-8413-a5fd4b4787a2.png">

It seems that when a device is running, it has a command called `avd` which can be run to get the name (`avd name`). So I don't think we can get the emulator name from the ACD, however I think in `flutter devices` we may be able to do the opposite.

I'm not sure if `flutter devices` already has the necessary plumbing to execute commands like this on the device (I'd guess so?). I'll do some digging after lunch and maybe discuss with @devoncarew during our call later.

DanTup commented:
@mit-mit How did you get that dialog? When I hit Run in AS with no emulators running, I get an error saying I must connect a device. If I have one running, it just launches on that directly. I'm not sure how to get the dialogs you showed?

I can find a few references online to the `avd name` command being available if you connect to the device, but it doesn't seem to work for me (suggests there's no `avd` command - so I'm curious whether the dialog you showed would work on my image).

mit-mit commented:
There is a setting for this in the Launch Configuration. Make sure Target is set to 'Open Select Deployment Target Dialog'.

DanTup commented:
Strange, I seem to have that set:

<img width="501" alt="screen shot 2018-04-19 at 2 13 23 pm" src="https://user-images.githubusercontent.com/1078012/38993565-ed3f90fc-43db-11e8-80ba-1f614e15b206.png">

Yet clicking Run or Debug gives this:

<img width="424" alt="screen shot 2018-04-19 at 2 13 47 pm" src="https://user-images.githubusercontent.com/1078012/38993582-f61d3828-43db-11e8-8d25-e68efa834c71.png">

If I start the emulator first, then the drop-down at the top changes to show it and then it always launches on that device. You're launching be pressing the Play/Debug buttons on the toolbar, right?

Is there a way to open the "Select deployment target" dialog without clicking run?

DanTup commented:
Aha, gotcha! Got it working now, and it is indeed working as you described. I'll see if I can figure out how it's matching these up. Thanks!

DanTup commented:
Ok, go tit:

```text
dantup-macbookpro:inetutils-1.9 dantup$ flutter emulators --start danny
dantup-macbookpro:inetutils-1.9 dantup$ adb devices
List of devices attached
emulator-5554	device

dantup-macbookpro:inetutils-1.9 dantup$ telnet/telnet localhost 5554
Trying ::1...
Connection failed: Connection refused
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
Android Console: Authentication required
Android Console: type 'auth <auth_token>' to authenticate
Android Console: you can find your <auth_token> in 
'/Users/dantup/.emulator_console_auth_token'
OK
avd name
Danny_Nexus
OK
```

I can't find a way to get this through `adb shell`, seems to be a special command that works over telnet. So I think whenever `flutter devices` polls, we will need to open a telnet connection, run `avd name` and then close the connection. I don't know how much more expensive this'll make it but it'll mean we could subtract these names from the list of AVDs and then add what's left to the end of the devices list and allow a user to launch them.

DanTup committed:
Re-order columns

devoncarew commented:
Taking a look at this again (and, sorry for the long delay!).

I have two comments on the UX, and will take a spin through the code after:

I'd like to keep lists of emulators out of the output of `flutter devices`. I don't have any super defensible reasons for this, but like the separation. Instead, if no devices are found in `flutter devices`, we should change our ending message in `flutter devices` from:

```
If you expected your device to be detected, please run "flutter doctor" to diagnose
potential issues, or visit https://flutter.io/setup/ for troubleshooting tips.
```

to something like:

```
Run 'flutter emulators' to list and start any available device emulators.

Or, if you expected your device to be detected, run "flutter doctor" to diagnose
potential issues or visit https://flutter.io/setup/ for troubleshooting tips.
```

The other UI think is that we could consider a different name for `--start`; perhaps `--launch`? `--run` might be conflated with actually starting an app.

devoncarew commented on packages/flutter_tools/lib/src/ios/ios_emulators.dart:
```diff
@@ -0,0 +1,71 @@
+
+// TODO(dantup): Is there a better name for this? We already have IOSSimulator classes
+// that represent *running* simulators, but this is about "unlaunched images"...
+class IOSEmulators extends EmulatorDiscovery {
```
IOSSimulatorDiscovery?

DanTup replied:
I think using the word Simulators is confusing, as we already use that for *running* devices (eg. we have a `class IOSSimulators extends PollingDeviceDiscovery`.

For Android we have "devices" (running) and "emulators" (AVDs), but for iOS we already had "devices" (running hardware) and "simulators" (running simulators). I think keeping it as "emulators" probably is least confusing for now; but probably it'd have been better if the existing "simulators" was named something less ambiguous (since we can have both offline and online simulators, whereas for devices we call the offline ones emulators - though that's still also kinda confusing).

I think probably I should just remove the TODO for now since I don't have any better idea (the more correct names are more ambiguous/confusing).

devoncarew replied:
We may need to differentiate between a running simulator, and a class which can start simulators. In any case, it's not a major issue; keeping the name as is and removing the TODO sounds good.

devoncarew commented on packages/flutter_tools/lib/src/ios/ios_emulators.dart:
```diff
@@ -0,0 +1,71 @@
+String getSimulatorPath() {
+  final List<String> searchPaths = <String>[
+    // TODO(dantup): Could this be anywhere else?
+    '/Applications/Xcode.app/Contents/Developer/Applications/Simulator.app',
```
@xster @cbracken would know

cbracken replied:
It could technically be anywhere at all :) The best way to locate which Xcode installation is currently in use is `xcode-select -p`. You can get the correct value from `xcode.xcodeSelectPath` defined in mac.dart.

DanTup replied:
Aha; swapped to that - thanks!

devoncarew commented on packages/flutter_tools/test/emulator_test.dart:
```diff
@@ -0,0 +1,70 @@
+  void launch() {
+    throw new UnimplementedError('Not implemented in Mock');
+  }
+
```
nit: extra whitespace

devoncarew reviewed (commented):
Looks good!

- I think we should update the exit message for `flutter devices` when there are no devices found, to reference `flutter emulators`
- can you clarify re: 'Issues with awaiting the launch of emulator' - I didn't find the reference in the PR
- as you mentioned in the PR, getting an integration test for this would be great. I think we can land without, and then work with @yjbanov to make sure that the test could run well on the device lab (and/or wait until we support `emulators create`, to ensure we have a device)

devoncarew reviewed (approved):
Thinking about it, we can land this w/o an update to the `flutter devices` no device message, and always incrementally improve that command in a later PR.

DanTup committed:
Update "no devices" text to suggest "flutter emulators"

DanTup committed:
Change start->launch when refering to emulators

DanTup commented:
@devoncarew 

> can you clarify re: 'Issues with awaiting the launch of emulator' - I didn't find the reference in the PR

Currently emulator.dart declares launch as returning `void`, but the overrides return Futures. If we change it to Future and then await it, then `flutter emulators --launch x` will await the completion of that process (wait until the user closes it).

If I don't await the process in the launch method, then I can't check the exit code. So, I guess I need to wait a short period (a second?) and if the process is still running, then assume it's a running emulator and *not* wait for it (we want to return the user to a terminal). I can't just fire-and-forget from the start because if it fails we'll give the user no indication of what it didn't launch (eg. not route the stderr to them).

DanTup committed:
Remove TODO

I can't come up with a better name; anything with Simulator or Device in it will be confused with the existing IOSSimulator/Device classes (which represent the running devices).

DanTup committed:
Remove superflous whitespace

DanTup committed:
Switch streams to Future<Lists>

DanTup committed:
Add missing await

devoncarew commented:
> If I don't await the process in the launch method, then I can't check the exit code. So, I guess I need to wait a short period (a second?) and if the process is still running, then assume it's a running emulator and not wait for it (we want to return the user to a terminal). I can't just fire-and-forget from the start because if it fails we'll give the user no indication of what it didn't launch (eg. not route the stderr to them).

Ah, gotcha. There's no signal to indicate a successful launch.

> Should we support flutter-tester here?

I'm not sure it helps, but haven't thought about it.

lgtm w/ the logic to check for an unsuccessful launch (that takes a few seconds to determine).

DanTup committed:
Add missing awaits

DanTup committed:
Only wait up to 3 seconds for emulator

Emaultor keeps running on a seuccessful launch, so this automatically returns after 3 seconds if the process hasn't quit (we have to wait for some period to get stderr in the case of a failure).

DanTup commented:
> I'm not sure it helps, but haven't thought about it.

I was thinking if `flutter-tester` was listed and launchable, it might give us a way to have some tests that don't don't can run outside of devicelab and don't require some specific setup on devicelab. However, it occurs to me now that `flutter devices` includes the test device when passed the flag always; it doesn't list it only when it's running (since it doesn't really run like a normal device that can be connected to).

Anyway, I've pushed a change to add the missing awaits and to automatically return from `AndroidEmulator.launch()` after 3 seconds if the process is still running.

There are two outstanding issues at the top; one is about Simulator location - I think we should probably just remove the TODO and assume the path is always this. If it's not, the worst to happen is that the simulator won't appear in the devices list.

The other one is about tests. I think the options are:

1. Leave them until we've done `--create` and then create devicelab tests that create, list, launch, verify launch
2. Ensure devicelab has some consistent AVDs (and iOS Simulator?) we can use

If we're going to do 1, it'd be good to have someone else check the code out and have a little test for sanity. LMK what you think!

DanTup committed:
Fix signature of mock launch

cbracken commented on packages/flutter_tools/lib/src/ios/ios_emulators.dart:
```diff
@@ -0,0 +1,66 @@
+import '../globals.dart';
+import 'ios_workflow.dart';
+
+class IOSEmulators extends EmulatorDiscovery {
```
nit: technically, iOS doesn't provide emulators, just simulators that run x86_64 code on a faked-out platform.

DanTup replied:
Yeah; this naming is something we couldn't really settle on. We already have a bunch of classes using the name "Simulator" which are used in the `devices` code (which represent active running simulators). The `emulators` section lists emulators/simulators that are not necessarily running. Naming these ones Simulators will cause confusion (or collisions) with the other classes (the ones that represent online simulators).

I'm happy to rename, but we couldn't come up with something more accurate that might not also be confusing. (I guess the same comment applies to the user-facing `flutter emulators` command too).

cbracken replied:
I'll be honest, I'm pretty stumped too. SimulatedDevice sounds too close to Simulator and technically it's an EmulatedDevice for Android. SoftwareDevice is pretty cumbersome and I don't think I've ever heard anyone use that term either way :) `IOSEmulators` seems like a reasonable compromise.

devoncarew commented:
I patched it in locally and gave it a spin:

<img width="477" alt="screen shot 2018-05-08 at 11 19 04 pm" src="https://user-images.githubusercontent.com/1269969/39798696-6db82868-5316-11e8-8e0c-40ca5c5f323f.png">

It looks great! We may want to tweak two things:

Above, there's a trailing bullet for the ios simulator. If no label for an emulator (like for simulators), we could not print the last bullet.

After printing the emulators description, we should give the user an indication for how to run one. So, if one or more emulators are found, print `To run an emulator, run 'flutter emulator --launch=<emulator id>` or similar.

DanTup committed:
Add 'emulator' as an alias for 'emulators'

It feels more nautral to type when doing `--launch x` or `--create x`.

DanTup committed:
Add text explaining how to run an emulator

DanTup committed:
Fix formatting

DanTup committed:
Don't show last bullet if there's no label

DanTup committed:
Use correct path for Simulator

DanTup commented:
> If no label for an emulator (like for simulators), we could not print the last bullet.

Done!

> After printing the emulators description, we should give the user an indication for how to run one.

Done!

> `flutter emulator --launch`

The command is actually `emulators` but I've added an alias since `emulator` feels more natural to type with `--launch` or `--create`, but I've used `emulators` (the way it's written in the help output) in the text.

DanTup commented:
I've pushed some changes; I think this covers all comments and outstanding issues with the exception of integration tests. @devoncarew lmk if you're happy enough with it to land as-is and put tests in once we have creation.

I'll give it a quick run on Windows again shortly with the latest changes too.

DanTup committed:
Trim all space/dots from end

Depending on whether can read the AVD (or it even has all fields populated) we might get extra "empty" columns, so this trims all blank cells from the end.

devoncarew commented:
👍 lgtm
